src/kernels/activation_functions.hpp

```diff
@@ -0,0 +1,546 @@
+                                                        UNUSED const FP_TYPE_PREC alpha)
+{
+    for(uint i = 0; i < n; ++i)
+    {
+        //	y = log(1 + exp(x))
+        res[i] = (data[i] > 0) ? (data[i] + log(static_cast<FP_TYPE_PREC>(1) + exp(-data[i])))
+                               : log(static_cast<FP_TYPE_PREC>(1) + exp(data[i]));
+    }
+}
+
+__forceinline__ __device__ void ActivationFunction_Leaky_ReLU(const uint n,
+                                                              FP_TYPE_PREC* res,
+                                                              const FP_TYPE_PREC* data,
+                                                              UNUSED const FP_TYPE_PREC gamma,
+                                                              UNUSED const FP_TYPE_PREC beta,
+                                                              const FP_TYPE_PREC alpha)
+{
+    for(uint i = 0; i < n; ++i)
+    {
+        res[i] = data[i] * ((data[i] > 0) ? static_cast<FP_TYPE_PREC>(1) : alpha);
+    }
+}
+
+__forceinline__ __device__ void ActivationFunction_Clipped_ReLU(const uint n,
+                                                                FP_TYPE_PREC* res,
+                                                                const FP_TYPE_PREC* data,
+                                                                UNUSED const FP_TYPE_PREC gamma,
+                                                                UNUSED const FP_TYPE_PREC beta,
+                                                                const FP_TYPE_PREC alpha)
+{
+    for(uint i = 0; i < n; ++i)
+    {
+        res[i] =
+            fmin(static_cast<FP_TYPE_PREC>(alpha), fmax(static_cast<FP_TYPE_PREC>(data[i]), 0));
+    }
+}
+
+__forceinline__ __device__ void ActivationFunction_ELU(const uint n,
+                                                       FP_TYPE_PREC* res,
+                                                       const FP_TYPE_PREC* data,
+                                                       UNUSED const FP_TYPE_PREC gamma,
+                                                       UNUSED const FP_TYPE_PREC beta,
+                                                       const FP_TYPE_PREC alpha)
+{
+    for(uint i = 0; i < n; ++i)
+    {
+        res[i] = (data[i] > 0) ? data[i] : (alpha * (exp(data[i]) - static_cast<FP_TYPE_PREC>(1)));
+    }
+}
+
+__forceinline__ __device__ void ActivationFunction(const uint n,
+                                                   FP_TYPE_PREC* res,
+                                                   const FP_TYPE_PREC* data,
+                                                   const FP_TYPE_PREC gamma,
+                                                   const FP_TYPE_PREC beta,
+                                                   const FP_TYPE_PREC alpha)
+{
+    if constexpr(MIOPEN_NRN_OP_ID == MIOPEN_NEURON_PASTHRU)
+    {
+        ActivationFunction_PassThru(n, res, data, gamma, beta, alpha);
+    }
```

Comment:
You can do such a trick in C++:
```C++
template<typename T, size_t N>
__forceinline__ __device__ void ActivationFunction_ELU(T (& __restrict__ res)[N],
                                                       const T (& __restrict__ data)[N],
                                                       UNUSED const T gamma,
                                                       UNUSED const T beta,
                                                       const T alpha)
{
    for(size_t i = 0; i < N; ++i)
    {
        res[i] = (data[i] > 0) ? data[i] : (alpha * (exp(data[i]) - static_cast<FP_TYPE_PREC>(1)));
    }
}

template<typename T, size_t N>
__forceinline__ __device__ void ActivationFunction(T (& __restrict__ res)[N],
                                                   const T (& __restrict__ data)[N],
                                                   const T gamma,
                                                   const T beta,
                                                   const T alpha)
{
    static_assert(N<=4);
    if constexpr(MIOPEN_NRN_OP_ID == MIOPEN_NEURON_PASTHRU)
    {
        ActivationFunction_PassThru(res, data, gamma, beta, alpha);
    }
......
```

A calling this function will be like this:
```C++
ActivationFunction(actRes, bnRes, gamma, beta, alpha);
```
Without explicitly passing `MIOPEN_READ_UNIT` and with guarantees that `actRes` and `bnRes` are equal and there will be no out of bound access ever.

Reply:
Very interesting! Good to know, thanks!